_posts/2017-11-09-totw-49.md

```diff
@@ -0,0 +1,285 @@
+---
+title: "Tip of the Week #49: Argument-Dependent Lookup"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/49
+type: markdown
+order: "049"
+---
+
+Originally posted as totw/49 on 2013-07-14
+
+*"...whatever disappearing trail of its legalistic argle-bargle one chooses to
+follow..." --Antonin Scalia, [U.S. v Windsor dissenting opinion]
+(http://www.supremecourt.gov/opinions/12pdf/12-307_6j37.pdf)**
+
+## Overview
+
+A function call expression such as `func(a,b,c)`, in which the function is named
+without the `::` scope operator, is called unqualified. When C++ code refers to
+a function by an unqualified name, the compiler performs a search for a matching
+function declaration. What is surprising to some people (and different from
+other languages) is that in addition to the caller's lexical scope, the set of
+search scopes is augmented by namespaces associated with the function argument
+types. This additional lookup is called Argument-Dependent Lookup (ADL). It's
+definitely happening in your code, so you'll be much better off with a basic
+understanding of how it works.
+
+## Name Lookup Basics
+
+A function call must be mapped to a single function definition by the compiler.
+This matching is done in two independent serial processing stages. First, *name
+lookup* applies some scope searching rules to produce a set of overloads
+matching the name of the function. Then *overload resolution* takes those
+overloads produced by that name lookup and tries to choose a best match for the
+arguments given at the call site. Keep this distinction in mind. Name lookup
+comes first, and it doesn't try to make any determination as to whether a
+function is a good match or not. It doesn't even consider the argument count. It
+just searches scopes for a function name. Overload resolution is a complex topic
+in its own right, but it's not our focus right now. Just know that it's a
+separate processing stage that gets its inputs from name lookup.
+
+When an unqualified function call is encountered, several independent search
+sequences can occur for that function name, each attempting to match the name to
+a set of overloads. The most obvious search is the outward search starting from
+the lexical scope of the call site:
+
+```c++
+    namespace b {
+    void func();
```

Comment:
Your call, but I'd like to standardize on indenting these for scope using 2 spaces. I'd suggest the standard 4, but the horizontal screen real-estate is rather precious here.

That said, I don't understand why this tip never had any indentation whatsoever. It certainly is useful to read the scopes.

Reply:
Namespaces don't get indented (for precisely the reason that horizontal real-estate is valuable).